File: src/api/dcps/java5/common/java/code/org/omg/dds/topic/TopicDescription.java
/* Copyright 2010, Object Management Group, Inc.
 * Copyright 2010, PrismTech, Inc.
 * Copyright 2010, Real-Time Innovations, Inc.
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.omg.dds.topic;

import java.io.Closeable;

import org.omg.dds.core.DDSObject;
import org.omg.dds.domain.DomainParticipant;
import org.omg.dds.type.TypeSupport;


/**
 * This interface is the base for {@link org.omg.dds.topic.Topic}, {@link org.omg.dds.topic.ContentFilteredTopic},
 * and {@link org.omg.dds.topic.MultiTopic}.
 * <p>
 * TopicDescription represents the fact that both publications and
 * subscriptions are tied to a single data type. Its attribute typeName
 * defines a unique resulting type for the publication or the subscription
 * and therefore creates an implicit association with a {@link org.omg.dds.type.TypeSupport}.
 * TopicDescription has also a name that allows it to be retrieved locally.
 *
 * @param <TYPE>    The concrete type of the data that will be published and/
 *                  or subscribed by the readers and writers that use this
 *                  topic description.
 */
public interface TopicDescription<TYPE> extends Closeable, DDSObject
{
    /**
     * Returns the {@link org.omg.dds.type.TypeSupport} used to create this TopicDescription.
     * @return  the {@link org.omg.dds.type.TypeSupport}.
     */
    public TypeSupport<TYPE> getTypeSupport();

    /**
     * Cast this topic description to the given type, or throw an exception if
     * the cast fails.
     *
     * @param <OTHER>   The type of the data exchanged on this topic,
     *                  according to the caller.
     * @return          this topic description
     * @throws          ClassCastException if the cast fails
     */
    public <OTHER> TopicDescription<OTHER> cast();

    /**
     * Returns the type name used to create the TopicDescription.
     * @return  the type name
     */
    public String getTypeName();

    /**
     * Returns the name used to create the TopicDescription.
     * @return  the name.
     */
    public String getName();

    /**
     * Returns the {@link org.omg.dds.domain.DomainParticipant} to which the TopicDescription belongs.
     * @return  the {@link org.omg.dds.domain.DomainParticipant}
     */
    public DomainParticipant getParent();

    /**
     * Dispose the resources held by this object.
     * <p>
     * A TopicDescription cannot be closed if it is in use by any
     * {@link org.omg.dds.pub.DataWriter}s or {@link org.omg.dds.sub.DataReader}s. With respect to
     * {@link org.omg.dds.topic.Topic}s specifically: a Topic cannot be closed if it has any
     * remaining {@link org.omg.dds.topic.ContentFilteredTopic}s or {@link org.omg.dds.topic.MultiTopic}s related
     * to it.
     *
     * @see     org.omg.dds.core.Entity#close()
     */
    @Override
    public void close();
}

File: src/api/dcps/c++/common/code/TopicDescription.cpp
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
#include "TopicDescription.h"

#include "os_report.h"
#include "u_user.h"
#include "DomainParticipant.h"
#include "TypeSupportMetaHolder.h"
#include "ReportUtils.h"

DDS::OpenSplice::TopicDescription::TopicDescription() :
    tsMetaHolder(NULL),
    nrUsers(0),
    participant(NULL)
{
    // topic_name, type_name and expression are of type String_var and therefore do not need
    // explicit initialization, since the constructor of the String_var already takes care of that.
}

DDS::ReturnCode_t
DDS::OpenSplice::TopicDescription::nlReq_init(
    DDS::OpenSplice::DomainParticipant *participant,
    const DDS::Char *topic_name,
    const DDS::Char *type_name,
    const DDS::Char *expression,
    DDS::OpenSplice::TypeSupportMetaHolder *ts_meta_holder)
{
    assert (participant != NULL);
    assert (topic_name != NULL);
    assert (type_name != NULL);
    assert (expression != NULL);

    this->topic_name = DDS::string_dup(topic_name);
    this->type_name  = DDS::string_dup(type_name);
    this->expression = DDS::string_dup(expression);

    (void) DDS::DomainParticipant::_duplicate(participant);
    this->participant = participant;

    if (ts_meta_holder != NULL) {
        (void) DDS::Object::_duplicate(ts_meta_holder);
        this->tsMetaHolder = ts_meta_holder;
    }

    return DDS::RETCODE_OK;
}

DDS::ReturnCode_t
DDS::OpenSplice::TopicDescription::wlReq_deinit()
{
    DDS::ReturnCode_t result = DDS::RETCODE_OK;

    if (this->nrUsers != 0) {
        result = DDS::RETCODE_PRECONDITION_NOT_MET;
        CPP_REPORT(result, "TopicDescription still in use.");
    } else {
        DDS::release(this->participant);
        this->participant = NULL;

        if (this->tsMetaHolder != NULL) {
            DDS::release(this->tsMetaHolder);
            this->tsMetaHolder = NULL;
        }
    }

    return result;
}


DDS::OpenSplice::TopicDescription::~TopicDescription()
{
    // Delegate to parent.
}

char*
DDS::OpenSplice::TopicDescription::get_type_name()
{
    DDS::ReturnCode_t result;
    char* name = NULL;

    CPP_REPORT_STACK();

    result = this->check();
    if (result == DDS::RETCODE_OK) {
        if (this->type_name != NULL ) {
            name = DDS::string_dup(this->type_name);
        }
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return name;
}

char*
DDS::OpenSplice::TopicDescription::get_name()
{
    DDS::ReturnCode_t result;
    char* name = NULL;

    CPP_REPORT_STACK();

    result = this->check();
    if (result == DDS::RETCODE_OK) {
        if (this->topic_name != NULL ) {
            name = DDS::string_dup(this->topic_name);
        }
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return name;
}

DDS::DomainParticipant_ptr
DDS::OpenSplice::TopicDescription::get_participant()
{
    DDS::ReturnCode_t result;
    DDS::DomainParticipant_ptr participant = NULL;

    CPP_REPORT_STACK();

    result = this->check();
    if (result == DDS::RETCODE_OK) {
        participant = DDS::DomainParticipant::_duplicate(this->participant);
    }

    CPP_REPORT_FLUSH(this, result != DDS::RETCODE_OK);

    return participant;
}

DDS::OpenSplice::TypeSupportMetaHolder*
DDS::OpenSplice::TopicDescription::get_typesupport_meta_holder()
{
    DDS::ReturnCode_t result;
    DDS::OpenSplice::TypeSupportMetaHolder *meta_holder = NULL;

    result = this->write_lock();
    if (result == DDS::RETCODE_OK) {
        if (this->tsMetaHolder == NULL) {
            /* The only way that tsMetaHolder == NULL here, is when the topic
             * proxy is obtained through find_topic() on the participant. */
            result = this->participant->nlReq_find_type_support_meta_holder(this->type_name, this->tsMetaHolder);
        }
        if (this->tsMetaHolder != NULL) {
            meta_holder = this->tsMetaHolder;
            DDS::Object::_duplicate(meta_holder);
        }
        this->unlock();
    }

    return meta_holder;
}

const char *
DDS::OpenSplice::TopicDescription::rlReq_get_topic_expression ()
{
    return this->expression;
}

void
DDS::OpenSplice::TopicDescription::wlReq_incrNrUsers()
{
    this->nrUsers++;
}

void
DDS::OpenSplice::TopicDescription::wlReq_decrNrUsers()
{
    this->nrUsers--;
}

::DDS::Long
DDS::OpenSplice::TopicDescription::rlReq_getNrUsers()
{
    return this->nrUsers;
}

const char *
DDS::OpenSplice::TopicDescription::rlReq_getName()
{
    return this->topic_name;
}
